application/cms/scanner_service.py

```diff
@@ -0,0 +1,46 @@
+import json
+
+import requests
+
+from application.cms.exceptions import UploadCheckPending, UploadCheckFailed, UploadCheckVirusFound
+from application.cms.service import Service
+
+
+class ScannerService(Service):
+    def init_app(self, app):
+        super().init_app(app)
+        self.base_url = self.app.config["ATTACHMENT_SCANNER_URL"]
+        self.token = self.app.config["ATTACHMENT_SCANNER_API_KEY"]
+        self.enabled = self.app.config["ATTACHMENT_SCANNER_ENABLED"]
+
+    def scan_file(self, filename, fileobj) -> bool:
+        """
+        return: True if scanned and safe; False if not scanned
+        raises: child of UploadException if scanned and a problem occurred
+        """
+        if self.enabled:
+            response = requests.post(
+                f"{self.base_url}/requests", headers={"Authorization": f"Bearer {self.token}"}, files={"file": fileobj}
+            )
+
+            status = response.json()["status"].lower()
+
+            if status == "ok":
+                return True
+            elif status == "pending":
+                self.logger.warning(f"Upload scan pending for `{filename}`: check back for result later")
+                raise UploadCheckPending("Upload check did not complete (pending)")
+            elif status == "failed":
+                self.logger.error(f"Upload scan failed for `{filename}`: {response.json()}")
+                raise UploadCheckFailed("Upload check could not be completed (an error occurred)")
+            elif status == "found":
+                self.logger.error(f"Upload scan detected a virus in `{filename}`: {response.json()}")
+                raise UploadCheckVirusFound("Virus scan has found something suspicious")
+
+        else:
+            self.logger.warning(f"File upload scanning disabled: writing `{filename}` without virus check")
+
+        return False
+
+
+scanner_service = ScannerService()
```

Comment:
I'm not a fan of the singleton pattern, but that seems to be how our existing services work, so... 🤷‍♂️

Reply:
Yeah, this is for the sake of consistency. Redoing how our services are initialised is a bigger thing, and one that I've thrown in the tech improvements board already.